### DISCUSSION TITLE
[Question] How to keep highlight when converting pdf to image

### DISCUSSION THREAD

**Discussion by CarolinaRDC** [pymupdf/PyMuPDF#1413] [Q&A] [answered]
**Configuration**

```
3.7.6 (tags/v3.7.6:43364a7ae0, Dec 19 2019, 00:42:30) [MSC v.1916 64 bit (AMD64)] 
 win32 
 
PyMuPDF 1.19.1: Python bindings for the MuPDF 1.19.0 library.
Version date: 2021-10-23 00:00:10.
Built for Python 3.7 on win32 (64-bit).
```

Hello, I've realised that once I've done the highlight annotation in the pdf I cannot keep it if I want to save it as an image. It automatically disappears. Is there a way of doing so?

The code:
```

def pdf_to_image(pdf_filename, x, highlight_text=None):

    if highlight_text is not None:
        highlight_within_pdf(pdf_filename, x, text_to_highlight=highlight_text, color=color['amarillo'])

        new_doc = fitz.open(REL["crops"] + '/' + 'new_crop_highlighted_{}.pdf'.format(x))
        new_page = new_doc.load_page(0)
        pix = new_page.get_pixmap(matrix=fitz.Matrix(150 / 72, 150 / 72), alpha=False)
        pix.save(REL["crops"] + '/' + 'cropped_page_{}.jpg'.format(x))
        new_doc.close()
```

```

def highlight_within_pdf(pdf_filename, x, text_to_highlight, color=None):
    doc = fitz.open(pdf_filename)
    page = doc.load_page(0)
    text_instances = page.search_for(text_to_highlight)

    highlight = page.add_highlight_annot(text_instances)
    highlight.set_colors({"stroke": color})
    highlight.update()
    doc.reload_page(page)

    doc.save(REL["crops"] + '/' + 'new_crop_highlighted_{}.pdf'.format(x), garbage=4, deflate=True, clean=True)
    doc.close()

```
There is no error. It just doesn't keep the highlighted sentence.
Thank you very much in advance

**Comment by JorjMcKie** [comment, marked as answer]
Page pixmap creation by default is **_with annotations included_**. So you should see them.
The only possible hickup is that right after an `annot.update()` MuPDF won't always have become aware of the changes PyMuPDF has inserted in the annotation.
So it is best to reload the page **_before taking the pixmap_**: `page = doc.reload_page(page)`.
But after going to a different page, or after opening the PDF after such changes will do the same job.
If you save the document, all outstanding changes should be saved as well.

BTW: If you save a pixmap, you cannot directly use JPEG as output format: MuPDF doen't support this. But you could use `pil_save()` instead.

But the following does work as desired:
```python
>>> doc=fitz.open("search.pdf")
>>> page=doc[0]
>>> annot = page.add_highlight_annot(page.search_for("beteigeuze"))
>>> annot.set_colors(stroke=(0,1,0))
>>> annot.update()
>>> pix =page.get_pixmap()
>>> pix.save("beteigeuze.png")
>>> doc.save("beteigeuze.pdf")
>>> 
```
Pixmap:
![grafik](https://user-images.githubusercontent.com/8290722/142466468-d3c90096-2070-442b-9206-09bd75f702ad.png)

The PDF looks the same.

**Reply by CarolinaRDC** [reply]
Exactly! I tried with PNG and it worked perfectly. 
I'm trying now with pil_save() because it is more suitable for our project to save it with JPEG output format.

Thank you very much for answering so so fast and for all the information given.

**Reply by JorjMcKie** [reply]
you are welcome 😎
